### DISCUSSION TITLE
mixing multiple wav files esp32 pcm5102

### DISCUSSION THREAD

**Discussion by drfhaust** [pschatzmann/arduino-audio-tools#868] [General]
good day i am a noob trying to make use of your library, I have been trying to make sense of it all. i am trying to play multiple wav files simultaneously and i have tried numerous things, below is my latest failure
`/**
 * @file player-sd-i2s.ino
 * @brief example using the SD library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioLibs/AudioSourceSD.h"

const char *startFilePath="/";
const char* ext="wav";
AudioSourceSD source(startFilePath, ext,5);
AudioSourceSD source1("/fo", ext,5);
I2SStream i2s;

WAVDecoder decoder;
WAVDecoder decoder1;
AudioPlayer player(source, i2s, decoder);
AudioPlayer player1(source1, i2s, decoder1);

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
 // AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  
    auto config = i2s.defaultConfig(TX_MODE);
  config.sample_rate = 44100; 
  config.bits_per_sample = 16;
  config.channels = 2;
  i2s.begin(config);

  // setup player
  //source.setFileFilter("*Bob Dylan*");
//player.setMetadataCallback(printMetaData);
  player.begin();
   player1.begin();
 
   
}

void loop() {
  player.copy();
   player1.copy();
}`

i was reading your documentation and you suggested an input mixer but i dont know how to implement that, can you help?

**Comment by drfhaust** [comment]
i have moved a step further, however i am still having issues, if one file ends before the other, then it stops playing totally and makes clicking sounds
`/**
 * @file streams-sd_mp3-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 file and output it on I2S
 * @version 0.1
 * @date 2021-96-25
 * 
 * @copyright Copyright (c) 2021 
 */


#include <SPI.h>
#include <SD.h>


#include "AudioTools.h"


#include "AudioCodecs/CodecWAV.h"


const int chipSelect=5;
I2SStream i2s; // final output of decoded stream
OutputMixer<int16_t> mixer(i2s, 2);                         // output mixer with 2 outputs mixing to AudioKitStream
WAVDecoder decoder;        // decode wav to pcm and send it to I2S
StreamCopy copier; 
File audioFile;
File audioFile1;
StreamCopy copier1(mixer, audioFile);                          // copies sound into mixer
StreamCopy copier2(mixer, audioFile1);                          // copies sound into mixer

void setup(){
  Serial.begin(115200);
 // AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup file
  SD.begin(chipSelect);
  audioFile = SD.open("/wavfile1.wav");
   audioFile1 = SD.open("/wavfile2.wav");

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  i2s.begin(config);

 mixer.begin();

  //copier.begin(i2s, mixer);

}

void loop(){
   copier1.copy();
  // write idx 1 to the mixer and flush (because stream count = 2)
  copier2.copy();

}`

**Comment by pschatzmann** [comment]
You need to reconfigure the mixer when one file ends or make sure that you still send some (empty) data.

**Comment by drfhaust** [comment]
please can you suggest an example? i barely made it this far. 

is there a function that can detect end of file for the wav decoder?

also where do i find the function to pass empty data to the copier

**Comment by pschatzmann** [comment]
I suggest that you look what the copy() call is returning: if there is no data copied you will get 0! calling available() on the file should also work...

**Reply by drfhaust** [reply]
please where can i find it, the function declearations for copier and mixer

**Reply by pschatzmann** [reply]
Best you use the Alphabetic Class documentation that you can find in the README.md.
Just search for Documentation...

**Comment by drfhaust** [comment]
hi phil, i have tried closing the fie based on your suggestion, and playing a silent wav file genetated from audacity but still no luck 


`/**
 * @file streams-sd_mp3-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 file and output it on I2S
 * @version 0.1
 * @date 2021-96-25
 * 
 * @copyright Copyright (c) 2021 
 */


#include <SPI.h>
#include <SD.h>


#include "AudioTools.h"


#include "AudioCodecs/CodecWAV.h"


const int chipSelect=5;
I2SStream i2s; // final output of decoded stream
OutputMixer<int16_t> mixer(i2s, 2);                         // output mixer with 2 outputs mixing to AudioKitStream
WAVDecoder decoder;        // decode wav to pcm and send it to I2S
StreamCopy copier; 
File audioFile;
File audioFile1;
StreamCopy copier1(mixer, audioFile);                          // copies sound into mixer
StreamCopy copier2(mixer, audioFile1);                          // copies sound into mixer

void setup(){
  Serial.begin(115200);
  //AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup file
  SD.begin(chipSelect);
  audioFile = SD.open("/wavfile1.wav");
   audioFile1 = SD.open("/wavfile2.wav");

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  i2s.begin(config);

 mixer.begin();

  //copier.begin(i2s, mixer);

}

void loop(){

 if (copier1.available()>0) {
    copier1.copy();
  } 

   else{
    copier1.end();
   

  } 
  
 if (copier2.available()>0) {
    copier2.copy();
  } 

   else  if (copier2.available()<1){
   audioFile1 = SD.open("/sil.wav");
   
    StreamCopy copier2(mixer, audioFile1);                          // copies sound into mixer
copier2.copy();
  }
  
   
  }`


heres what i get on serial
`
[E] StreamCopy.h : 289 - write to target has failed! (1 bytes)
[E] StreamCopy.h : 289 - write to target has failed! (1 bytes)
[E] StreamCopy.h : 289 - write to target has failed! (1 bytes)
`

**Comment by pschatzmann** [comment]
Messing around with multiple files is way too complicated, why don't just just write an array with 0 to the destination instead?
Alternatively did you try to reconfigure the MultiOutput by calling setOutputCount(1) when the the first file has been closed ? Not sure if you need to call begin again for the change to take effect.

By the way to have things absolutely correct you would need to decode the wav data instead of copying the files directly, or start the file at position 44 to ignore the WAV header

**Reply by drfhaust** [reply]
i tried using the wav decoder but i couldnt get it to fit into the mixer , kept throwing an error

**Reply by drfhaust** [reply]
i tried your suggestion but it crashed 
`assert failed: size_t audio_tools::OutputMixer<T>::write(int, const uint8_t*, size_t) [with T = short int; size_t = unsigned int; uint8_t = unsigned char] AudioOutput.h:374 (p_buffer != nullptr)


Backtrace: 0x4008393d:0x3ffb25f0 0x4008aff5:0x3ffb2610 0x400900ed:0x3ffb2630 0x400d4d8a:0x3ffb2760 0x400d4f45:0x3ffb2780 0x400d45bc:0x3ffb27a0 0x400d46e8:0x3ffb27c0 0x400d4c4c:0x3ffb2800 0x400d9025:0x3ffb2820




ELF file SHA256: cfee131e204ad119


`

**Comment by drfhaust** [comment]
any comment on this?
